nntrainer/layers/preprocess_translate_layer.cpp

```diff
@@ -0,0 +1,134 @@
+  output_dim = input_dim;
+
+  rng.seed(getSeed());
+
+  // Made for 3 channel input
+  if (translation_factor > epsilon) {
+    if (input_dim[0].channel() > 3)
+      throw exception::not_supported(
+        "Preprocess translate layer not supported for over 3 channels");
+    translate_dist = std::uniform_real_distribution<float>(-translation_factor,
+                                                           translation_factor);
+
+#if defined(ENABLE_DATA_AUGMENTATION_OPENCV)
+    affine_transform_mat = cv::Mat::zeros(2, 3, CV_32FC1);
+    affine_transform_mat.at<float>(0, 0) = 1;
+    affine_transform_mat.at<float>(1, 1) = 1;
+
+    input_mat =
+      cv::Mat::zeros(input_dim[0].height(), input_dim[0].width(), CV_32FC3);
+    output_mat =
+      cv::Mat::zeros(input_dim[0].height(), input_dim[0].width(), CV_32FC3);
+#else
+    throw exception::not_supported(
+      "Preprocess translate layer is not supported without opencv");
+#endif
+  }
+
+  return ML_ERROR_NONE;
+}
+
+void PreprocessTranslateLayer::setProperty(const PropertyType type,
+                                           const std::string &value) {
+  int status = ML_ERROR_NONE;
+
+  switch (type) {
+  case PropertyType::random_translate:
+    if (!value.empty()) {
+      status = setFloat(translation_factor, value);
+      translation_factor = std::abs(translation_factor);
+      throw_status(status);
+    }
+    break;
+  default:
+    Layer::setProperty(type, value);
+    break;
+  }
+}
+
+void PreprocessTranslateLayer::forwarding() {
+  for (unsigned int idx = 0; idx < input_dim.size(); idx++) {
+
+    Tensor &hidden_ = net_hidden[idx]->getVariableRef();
+    Tensor &input_ = net_input[idx]->getVariableRef();
+
+    if (translation_factor < epsilon) {
+      hidden_ = input_;
+      continue;
+    }
+
+#if defined(ENABLE_DATA_AUGMENTATION_OPENCV)
+    for (unsigned int b = 0; b < input_dim[idx].batch(); b++) {
+
+      /** random translation */
+      float translate_x = translate_dist(rng) * input_dim[idx].width();
+      float translate_y = translate_dist(rng) * input_dim[idx].height();
+      affine_transform_mat.at<cv::Vec2f>(0, 0)[2] = translate_x;
+      affine_transform_mat.at<cv::Vec2f>(1, 0)[2] = translate_y;
+
+      for (unsigned int c = 0; c < input_dim[idx].channel(); c++)
+        for (unsigned int h = 0; h < input_dim[idx].height(); h++)
+          for (unsigned int w = 0; w < input_dim[idx].width(); w++)
+            input_mat.at<cv::Vec3f>(h, w)[c] = input_.getValue(b, c, h, w);
+
+      cv::warpAffine(input_mat, output_mat, affine_transform_mat,
+                     output_mat.size(), cv::WARP_INVERSE_MAP,
+                     cv::BORDER_REFLECT);
+
+      for (unsigned int c = 0; c < input_dim[idx].channel(); c++)
+        for (unsigned int h = 0; h < input_dim[idx].height(); h++)
+          for (unsigned int w = 0; w < input_dim[idx].width(); w++)
+            hidden_.setValue(b, c, h, w, output_mat.at<cv::Vec3f>(h, w)[c]);
```

Comment:
Once we finish line # 103, we can use the buffer for line112. I guess.

Reply:
The bottleneck is line 105, where `input_mat` and `output_mat` have to be different.

Reply:
> The bottleneck is line 105, where `input_mat` and `output_mat` have to be different.
> We can use `input_` and `hidden_` to be the same for now, but it might lead to optimizations, as this layer cannot work in-place. It just seems to work in-place because of the extra memory allocated by OpenCV.

I'm saying for the opencv case.

Reply:
done.